### ISSUE TITLE
Any ways to supply custom ChanelPipeline?

### ISSUE THREAD

**Issue by Nashatyrev** [netty/netty#9505]
Hey Netty team,

Started to extend the `DefaultChannelPipeline` and found out that it is not possible due to: 
- `AbsractChannel` relies on `DefaultChannelPipeline` class rather than `ChannelPipeline` interface
- `DefaultChannelPipeline` is 'effectively final' (all interface methods are implemented as `final`)

I would suggest either one of the following options: 
1. Change `AbstractChannel` to depend on `ChannelPipeline` instead of `DefaultChannelPipeline` (the same as 'Idea 1' in #7484)
I've found just 2 usages of `DefaultChannelPipeline` specific methods: `estimatorHandle()` and `invokeHandlerAddedIfNeeded()` which can be pulled-up to the interface (with maybe slight redesign)
2. Relax `DefaultChannelPipeline` and remove final modifiers
3. At least add pipeline modification callbacks (like handlerAdded/Removed) to the `DefaultChannelPipeline`

Background use case: https://github.com/libp2p/jvm-libp2p/issues/39
Closely relates to #7484

**Comment by normanmaurer** [comment]
While I think in general this is a good idea I don't think we can make this happen with the suggestions you made without breaking the API. Which means this can not happen in 4.1.x

**Comment by Nashatyrev** [comment]
@normanmaurer thanks for prompt response!

May be you give any advice on how to track handlers add/remove with the current API?

**Comment by normanmaurer** [comment]
what exactly you need to track ? Just when a handler is added and removed ?

**Comment by Nashatyrev** [comment]
> Just when a handler is added and removed ?
Yep. 
The general idea is always add a tailing handler which should cache all events (which would be dropped by pipeline else) and replay them to a new handler upon addition

**Comment by njhill** [comment]
@Nashatyrev I didn't digest your problem in detail but might `ChannelPipeline::replace` help here?

**Comment by Nashatyrev** [comment]
@njhill Nope. I want to be notified on adding/removing handlers to pipeline

**Comment by normanmaurer** [comment]
@Nashatyrev I see... I mean what you could do is fire user events through the pipeline on each remove / add.

**Comment by normanmaurer** [comment]
We don't plan to add this
